repository/copier.go

```diff
@@ -4,107 +4,191 @@ import (
+	return c.local.Filesystem
 }
 
-func (c *LocalCopier) CopyFromRemote(src, dst string, localFs billy.Filesystem) error {
-	bSrc := addBucketName(src, c.bucketSize)
-	return c.copyFile(c.fs, localFs, bSrc, dst)
-}
+// CopyToRemote copies a file from the local filesystem to the remote one.
+func (c *Copier) CopyToRemote(ctx context.Context, src, dst string) error {
+	dstCopy := filepath.Join(c.remote.TempDir(), dst+".copy")
+	dst = filepath.Join(
+		c.remote.Base(),
+		addBucketName(dst, c.bucketSize),
+	)
 
-func (c *LocalCopier) CopyToRemote(src, dst string, localFs billy.Filesystem) error {
-	bDst := addBucketName(dst, c.bucketSize)
-	return c.copyFile(localFs, c.fs, src, bDst)
-}
+	lf, err := c.local.Open(filepath.Join(c.local.Base(), src))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
 
-func (c *LocalCopier) copyFile(fromFs, toFs billy.Filesystem, from, to string) (err error) {
-	src, err := fromFs.Open(from)
-	if os.IsNotExist(err) {
-		return nil
+	defer checkClose(lf, &err)
+
+	w, err := c.remote.WriteTo(dstCopy)
+	if err != nil {
+		return err
 	}
 
+	// Delete temporary file in case the process is stopped while copying
+	defer func() {
+		_ = w.Close()
+		_ = c.remote.DeleteIfExists(dstCopy)
+	}()
+
+	err = copy(ctx, w, lf)
+
+	checkClose(w, &err)
+
 	if err != nil {
+		_ = c.remote.DeleteIfExists(dstCopy)
 		return err
 	}
-	defer checkClose(src, &err)
 
-	dst, err := toFs.OpenFile(to, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0644))
+	err = c.remote.Rename(dstCopy, dst)
 	if err != nil {
 		return err
 	}
-	defer checkClose(dst, &err)
 
-	_, err = io.Copy(dst, src)
 	return err
 }
 
-// NewHDFSCopier returns a copier using as a remote an HDFS cluster.
+// CopyFromRemote copies a file to the local filesystem from the remote one.
+func (c *Copier) CopyFromRemote(ctx context.Context, src, dst string) (err error) {
+	src = filepath.Join(
+		c.remote.Base(),
+		addBucketName(src, c.bucketSize),
+	)
+	rf, err := c.remote.Open(src)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	defer checkClose(rf, &err)
+
+	w, err := c.local.WriteTo(filepath.Join(c.local.Base(), dst))
+	if err != nil {
+		return err
+	}
+	defer checkClose(w, &err)
+
+	err = copy(ctx, w, rf)
+	return
+}
+
+type localFs struct {
+	billy.Filesystem
+}
+
+// NewLocalFs returns a local file system.
+func NewLocalFs(fs billy.Filesystem) Fs {
+	return &localFs{fs}
+}
+
+func (fs *localFs) Open(path string) (io.ReadCloser, error) {
+	return fs.Filesystem.Open(path)
+}
+
+func (fs *localFs) WriteTo(path string) (io.WriteCloser, error) {
+	return fs.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0644))
+}
+
+func (fs *localFs) DeleteIfExists(path string) error {
+	if _, err := fs.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	return fs.Remove(path)
+}
+
+func (fs *localFs) TempDir() string {
+	return os.TempDir()
```

Comment:
Does it mean that when it's using local filesystem and is pushing a siva file it is first copied to `/tmp` and then renamed to the rooted repositories dir?

If this is true I would make this configurable as before or copy directly to the destination directory if you think it is too cumbersome. The idea is that the last rename is just a metadata change instead of a full copy of the contents. In most systems `/tmp` is not the same filesystem as the place where you are pushing the files and that would make the rename operation to do a full copy of its contents.

Reply:
Yep, the problem is no matter the fs, both share the exact same code. It could return `""` instead and let the user handle completely the paths in this case, I guess.

Reply:
Can we keep temporary name (just for safety) but copy to the destination dir, and after that just rename a file?

Reply:
In the previous version we copied to a different directory because the rooted repo became polluted with `.siva.copy` files. This caused complains when engine errored because of them or because when the rooted repo was recursively copied locally for other analysis they contained those files.

I believe that engine issue is solved and the other problem is easily solved with `rm */*.siva.copy`. That's why I say that you may copy directly to the destination directory if you find this too cumbersome and analyze later if this causes any issue.